README
Dist::Zilla::Plugin::MetaData::BuiltWith

This module provides extra metadata in your distribution, automatically
documenting what versions of dependencies the author was using at the
time of release.

This allows consumers of said distributions to be able to see a range of
versions that are "known good" should they experience problems.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Dist-Zilla-Plugin-MetaData-BuiltWith

1.004006 2017-03-02T03:21:14Z 13ad9b6

1.004005 2017-03-02T03:06:51Z 15987f6
 [BugFix]
 - Avoid test failures on 5.26 w/ -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/Dist-Zilla-Plugin-MetaData-BuiltWith/issues/2

 [Dependencies::Stats]
 - Dependencies changed since 1.004004, see misc/*.deps* for details
 - develop: +1 ↑1
 - test: ↓1

1.004004 2016-09-10T10:08:39Z d8d12b1
 [Bugfix]
 - Don't die when uname call fails.
 - To be honest, I'm still not sure why I was doing this in the first place.
 - Patch Patched (with thanks) by Roy Ivy II ( #1 )

 [Dependencies::Stats]
 - Dependencies changed since 1.004003, see misc/*.deps* for details
 - develop: (suggests: ↑1)

1.004003 2016-05-01T21:35:06Z e12fda1
 - Removed DZTest stuff.
 - Removed use of Util::ConfigDumper
 - Removed use of Module::Data & Path::ScanINC from the depgraph.

 [Dependencies::Stats]
 - Dependencies changed since 1.004002, see misc/*.deps* for details
 - develop: +5 ↑1 -1 (suggests: ↑2)
 - runtime: +2 -2 (recommends: +1)
 - test: -2 (recommends: -1)

1.004002 2015-03-06T23:08:04Z 81ea208
 [Dependencies::Stats]
 - Dependencies changed since 1.004001, see misc/*.deps* for details
 - configure: (recommends: ↓1)
 - develop: +3 ↑2 -1 (recommends: ↓1, suggests: ↑1)
 - runtime: (recommends: +1)
 - test: (recommends: +1 ↓2)

 [Internals]
 - Cleanup some residual cruft from de-autoboxing.

1.004001 2014-12-29T14:33:59Z 8a275b9
 [Bugfix]
 - Fix emitting 'null' for version in external file format.

 [Dependencies::Stats]
 - Dependencies changed since 1.004000, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 ↑1 (recommends: ↑1, suggests: ↑1)
 - runtime: ↓1 -1
 - test: ↓1 (recommends: ↑1)

 [Internals]
 - Slightly better code refactory in main loop.

1.004000 2014-12-14T08:34:41Z 24fbb48
 [Dependencies::Stats]
 - Dependencies changed since 1.003001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +6 ↑1 -2 (recommends: +1, suggests: ↑2)
 - runtime: +3
 - test: +1 (recommends: ↑2)

 [Feature]
 - Can now remove the data from META.* and export into a single file on its own using 'use_external_file = only'.
 - File will be a name of user choice but defaults to 'misc/built_with.json'

1.003001 2014-09-24T06:28:04Z
 [Bugfix]
 - Don't call unname on devices where there is no uname.

 [Dependencies::Stats]
 - Dependencies changed since 1.003000, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: +1
 - test: (recommends: ↑1)

1.003000 2014-09-03T22:13:41Z
 [00 Minor]
 - Internals.
 - Tests.

 [Bugfix]
 - BuiltWith::All could crash trying to load some invalid module names it erroneously thought existed by traversing the
   global stash.

 [Dependencies::Stats]
 - Dependencies changed since 1.002000, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: +2
 - test: +2

 [Internals]
 - dump_config impemented with config_dumper

 [Tests]
 - Tests reimplented in dztest

1.002000 2014-08-18T03:26:39Z
 [00 Minor]
 - use JSON::MaybeXS instead of JSON

 [Dependencies::Stats]
 - Dependencies changed since 1.001001, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - runtime: +1 -1

1.001001 2014-08-15T21:17:00Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests augmented.
 - whitespace adjusted.

 [Dependencies::Stats]
 - Dependencies changed since 1.001000, see misc/*.deps* for details
 - develop: +1 (suggests: ↑1)
 - test: -3

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update reportprereqs test

1.001000 2014-07-30T13:22:16Z
 [00 Major]
 - Added tests
 - Prereqs retrieved directly from Dist::Zilla now *WITHOUT* using EmulatePhase.
 - x_BuiltWith keys now added during META.* write out, which happens after prereqs finalization, by currying the
   FromCode coderef and injecting the data raw and de/reserializing.

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: ↓1
 - develop: (suggests: ↑1)
 - runtime: +4 -4
 - test: +2 -1 (recommends: ↑1 -2)

1.000001 2014-07-19T02:16:32Z
 [00 Trivial]
 - Misc packaging cleanups

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - develop: +3 -3 (suggests: +2 -1)
 - test: +2 ↑1 (recommends: +1)

1.000000 2014-05-20T05:33:17Z
 [00 Major]
 - Moderately sized packaging changes, a few small code changes.

 [01 Tooling]
 - Now uses ExtUtils::MakeMaker for installer.
 - "Latest" dependency logic now only added to /recommends/

 [02 Versioning]
 - Version scheme is now x.yyyyyy ( mantissa = 6 )
 - This is the primary reason for the 1.x on the box.

 [Bugfix]
 - BuiltWith::All ocasionally messed up when it found certain modules in %INC which had names such that it was
   impossible to load them ( Anon classes, Mx Types ), and these needlessly caused failures. Fixed.

 [Dependencies::Aggregate]
 - 5.008009 : 202 → 206
 - 5.010001 : 180 → 185
 - 5.018002 : ? → 172
 - 5.019000 : 166 → ?
 - 5.019009 : ? → 166

 [Dependencies::Stats]
 - Dependencies changed since 0.04000002, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +7 ↑1 -9
 - runtime: +1 ↑1
 - test: +4 ↓1 (recommends: +4)

0.04000002 2013-11-25T08:35:49Z
 [00 Bugfix]
 - Don't try to test if it was a file or not if its undefined.

 [Dependencies::Stats]
 - Dependencies changed since 0.04000001, see misc/*.deps* for details
 - develop: ↑1
 - runtime: +1

0.04000001 2013-11-22T04:23:48Z
 [00 Maint]
 - Changes retroactively cleaned. Next time, you will be forced to wear a clown shoes and ride a unicycle.

 [Dependencies::Stats]
 - Dependencies changed since 0.04000000, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +43 ↑1 -1 (recommends: -1, suggests: +1 -1)
 - test: ↑1 -3

 [Documentation]
 - Readme.mkdn → MetaCPAN

 [Packaging]
 - Build.PL updated by Dzil to have new Fallback Mechanics.

 [Tests]

Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4202"
  },
  "configure_requires" => {
    "Module::Build" => "0.4202"
  },
  "dist_abstract" => "Report what versions of things your distribution was built against",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-MetaData-BuiltWith",
  "dist_version" => "0.04000002",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Plugin::MetaData::BuiltWith",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Class::Load" => 0,
    "Dist::Zilla::Role::ConfigDumper" => "4.101540",
    "Dist::Zilla::Role::MetaProvider" => 0,
    "Dist::Zilla::Util::EmulatePhase" => "0.01000101",
    "Module::Data" => 0,
    "Moose" => "2.0",
    "MooseX::Types::Moose" => 0,
    "Readonly" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Test::More" => "1.001002"
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.4202",
  "Test::More" => "1.001002"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    metadata
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Report what versions of things your distribution was built against",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Plugin-MetaData-BuiltWith",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::Plugin::MetaData::BuiltWith",
  "PREREQ_PM" => {
    "CPAN::Meta::Converter" => 0,
    "Carp" => 0,
    "Devel::CheckBin" => 0,
    "Dist::Zilla::File::FromCode" => 0,
    "Dist::Zilla::Role::ConfigDumper" => "4.101540",
    "Dist::Zilla::Role::FileGatherer" => 0,
    "Dist::Zilla::Role::FileMunger" => 0,
    "Dist::Zilla::Role::MetaProvider" => 0,
    "JSON::MaybeXS" => 0,
    "Module::Metadata" => 0,
    "Module::Runtime" => 0,
    "Moose" => "2.0",
    "MooseX::Types::Moose" => 0,
    "Path::Tiny" => 0,
    "YAML::Tiny" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::DZil" => 0,
    "Test::More" => 0,
    "constant" => 0
  },
  "VERSION" => "1.004006",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Meta::Converter" => 0,
  "Carp" => 0,
  "Devel::CheckBin" => 0,
  "Dist::Zilla::File::FromCode" => 0,
  "Dist::Zilla::Role::ConfigDumper" => "4.101540",
  "Dist::Zilla::Role::FileGatherer" => 0,
  "Dist::Zilla::Role::FileMunger" => 0,
  "Dist::Zilla::Role::MetaProvider" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "JSON::MaybeXS" => 0,
  "Module::Metadata" => 0,
  "Module::Runtime" => 0,
  "Moose" => "2.0",
  "MooseX::Types::Moose" => 0,
  "Path::Tiny" => 0,
  "Test::DZil" => 0,
  "Test::More" => 0,
  "YAML::Tiny" => 0,
  "constant" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
